### PR TITLE
Grant Application HexPlay Project

### PR DESCRIPTION
**Pull request by HexPlayLab**
# Grant Application Checklist

- [x] The [application-template.md](https://github.com/w3f/Open-Grants-Program/blob/master/applications/application-template.md) has been copied, renamed ( "project_name.md") and updated.
- [x] A BTC address for the payment of the milestones is provided inside the application.  
- [x] The software of the project will be released under the Apache license version 2.0 as specified in the [terms and conditions](https://gist.github.com/Noc2/75bc58e8ce9b5d419ff883b0cf2b8c19).
- [x] The total funding amount of the project is below CHF 30k at the time of submission.
- [x] The initial PR contains only one commit (squash if needed before submitting your PR).

### CODE REVIEW

**Review by mmagician — CHANGES_REQUESTED** [review]
@HexPlayLab Thanks for another application.
1. Please indicate that you've applied for another [grant](https://github.com/w3f/Open-Grants-Program/pull/169) in the past.
2. Could you clarify the distinction between the fee for purchasing eggs in the game vs. fee for playing?
3. Milestone 1 doesn't provide any concrete substrate deliverables. The specification of your game & fleshing out the documentation is considered part of the application itself, rather than a milestone on its own & as such won't be explicitly funded. Please merge the M1 & M2.
4. Rather than implementing your own Swap module, would it make sense to perhaps re-use another project's work? (e.g. [subdex](https://github.com/subdarkdex/pallet-subdex/tree/master/pallet-subdex))
5. I think I understand the idea of the `Profit Pool` as described in the Project Details section, but I don't get how you'll implement it as a substrate module. Could you clarify what functionality will be available via this pallet? (have a look at the [requirement list](https://github.com/w3f/Open-Grants-Program/blob/master/src/grant_guidelines_per_category.md#requirements) and try to provide as much info as currently known for your modules)

**Comment by HexPlayLab** [comment]
> @HexPlayLab Thanks for another application.
> 
> 1. Please indicate that you've applied for another [grant](https://github.com/w3f/Open-Grants-Program/pull/169) in the past.
> 2. Could you clarify the distinction between the fee for purchasing eggs in the game vs. fee for playing?
> 3. Milestone 1 doesn't provide any concrete substrate deliverables. The specification of your game & fleshing out the documentation is considered part of the application itself, rather than a milestone on its own & as such won't be explicitly funded. Please merge the M1 & M2.
> 4. Rather than implementing your own Swap module, would it make sense to perhaps re-use another project's work? (e.g. [subdex](https://github.com/subdarkdex/pallet-subdex/tree/master/pallet-subdex))
> 5. I think I understand the idea of the `Profit Pool` as described in the Project Details section, but I don't get how you'll implement it as a substrate module. Could you clarify what functionality will be available via this pallet? (have a look at the [requirement list](https://github.com/w3f/Open-Grants-Program/blob/master/src/grant_guidelines_per_category.md#requirements) and try to provide as much info as currently known for your modules)

@mmagician Thanks for your comments.
1. The information about our previous application is added in the “Additional Information” section.
2. More details about the fee collected in a game is added in the “Game Center” section. Basic idea is that game needs development work to be done by some one (future update, routinely maintenance), thus games are encourage to be designed as somewhat profitable to cover its operation/maintenance expense. In the CyrptoZoo example, fee can be collected in two ways. First way is that when player purchases an egg, 5% of the money paid is collected. Secondly, when players buy an animal in the secondary market (in a client to client manner), 1% transaction fee is charged for both side. 
3. M1 and M2 are merged. 
4. Of course, reusing other people’s work would make sense. We may considering using subdex, zenLink for constructing our swap.
5. More details about the distribution of the profit pool are added in the profit pool section, and the changes are made to the milestone to clarify this confusion. Profit pool itself is not a module. Once the fee is collected, it is divided into two part. One part goes to the account of treasury, and the other part goes to the account of the mining pool. The LP of the swap locks LP tokens, and HPT holders locks HPT for mining the second part of the profit pool. Therefore the substrate module will be completed is this mining pool, where LP and HPT holders locked their token for mining profits of the platform.

**Comment by mmagician** [comment]
@HexPlayLab Thanks for the update, that makes it clearer.

> Milestone 1 doesn't provide any concrete substrate deliverables. The specification of your game & fleshing out the documentation is considered part of the application itself, rather than a milestone on its own & as such won't be explicitly funded. Please merge the M1 & M2.

What I meant by this is that we will not financially support the items from your original M1, but that this work should be part of your own effort & investment towards the grant. Furthermore, ideally some of it (your website, documentation drafts) could already be completed before receiving the grant. Therefore, I would ask you to keep the new deliverables, but fall back to 0.3 & 0.3 BTC per new milestones M1 & M2 (to discount the development of swap module in favour of using an existing one).

Lastly, could you please provide additional information on `Staking Dot` & `Mining Profit` modules that you plan to develop as part of M2? At the present state of information I don't quite know what to expect from these modules. Have a look at [this document](https://github.com/w3f/Open-Grants-Program/blob/master/src/grant_guidelines_per_category.md#requirements) for an idea of what we're looking for - together with a high-level description of functionality that is more detailed than what's already provided under the milestone.

**Comment by HexPlayLab** [comment]
> @HexPlayLab Thanks for the update, that makes it clearer.
> 
> > Milestone 1 doesn't provide any concrete substrate deliverables. The specification of your game & fleshing out the documentation is considered part of the application itself, rather than a milestone on its own & as such won't be explicitly funded. Please merge the M1 & M2.
> 
> What I meant by this is that we will not financially support the items from your original M1, but that this work should be part of your own effort & investment towards the grant. Furthermore, ideally some of it (your website, documentation drafts) could already be completed before receiving the grant. Therefore, I would ask you to keep the new deliverables, but fall back to 0.3 & 0.3 BTC per new milestones M1 & M2 (to discount the development of swap module in favour of using an existing one).
> 
> Lastly, could you please provide additional information on `Staking Dot` & `Mining Profit` modules that you plan to develop as part of M2? At the present state of information I don't quite know what to expect from these modules. Have a look at [this document](https://github.com/w3f/Open-Grants-Program/blob/master/src/grant_guidelines_per_category.md#requirements) for an idea of what we're looking for - together with a high-level description of functionality that is more detailed than what's already provided under the milestone.

Thanks for your followup. Our team is glad to see that the project makes sense to you. Regarding the comment you gave, we make the following changes.
(1) BTC amount is changed to 0.3, 0.3 per M1 and M2 respectively. 
(2) More details are added to the specification of staking Dot substrate module, and staking HPT/LP Token to mine profits module. Those details are basically what functions we aim to realize in the module, the high-level description is provided.
(3) Swap module development is removed from M2

Your thoughtful comments are valuable to us, and we are looking forward to your new response.

**Comment by mmagician** [comment]
Thank you for updating the document. I would be willing to support the project.

**Review by mmagician — DISMISSED** [review]

**Review by Noc2 — CHANGES_REQUESTED** [review]
Thanks for the application: I’m willing to support milestone 1, but I’m not sure about milestone 2. Why would you need to have your own “Substrate Module: Staking Dot” module? I guess you would need some kind of oracle for this to work. Because if you for example participate in the slot auction via [crowdfunding]( https://github.com/paritytech/polkadot/tree/master/runtime/common/src). Your independent chain doesn’t have any information about the status of the slot auction or the dot holdings. And something to be aware of is that tokens (dots/ksm) need to be unstaked in order to participate in a slot auction.

**Comment by HexPlayLab** [comment]
> Thanks for the application: I’m willing to support milestone 1, but I’m not sure about milestone 2. Why would you need to have your own “Substrate Module: Staking Dot” module? I guess you would need some kind of oracle for this to work. Because if you for example participate in the slot auction via [crowdfunding](https://github.com/paritytech/polkadot/tree/master/runtime/common/src). Your independent chain doesn’t have any information about the status of the slot auction or the dot holdings. And something to be aware of is that tokens (dots/ksm) need to be unstaked in order to participate in a slot auction.

Hello, thanks for your comments. There might be some confusion here caused by the inaccurate name of "staking dot module".  The functionality we try to realize in this module, is to snap-shoot which user deposit how much Dot, and later, if we succeed in securing a slot, distribute our native token to those user. That's all about this module. That has nothing to do with the standard interpretation of taking process as part of POS chain. Regarding the oracle, we currently doesn't intend to have one. We have a module to help raise Dot for auction, the Dot deposited to this module will be later transferred to the accout

**Comment by HexPlayLab** [comment]
> Thanks for the application: I’m willing to support milestone 1, but I’m not sure about milestone 2. Why would you need to have your own “Substrate Module: Staking Dot” module? I guess you would need some kind of oracle for this to work. Because if you for example participate in the slot auction via [crowdfunding](https://github.com/paritytech/polkadot/tree/master/runtime/common/src). Your independent chain doesn’t have any information about the status of the slot auction or the dot holdings. And something to be aware of is that tokens (dots/ksm) need to be unstaked in order to participate in a slot auction.

Thanks for your comment. That's really helpful. From your crowdfund link, we notice that there is a module available in the relay chain which helps us do the crowdfund in a safe and decentralized way. That's awesome. Therefore, we intend to take advantage of the crowdfund module to do the work. The original "staking dot module" is renamed to the "Distribution HPT module". In the distribution module, what we try to realize is to reward those depositing Dot into the campaign and helping us do the auction with our native token HPT. The exact distribution plan will be craft, and written into the logic of the sendReward() function.

The .md file is updated, and the changed part is the "staking module" -> "distribution module".

**Comment by Noc2** [comment]
Thanks for the update. Your [treasury proposal](https://docs.google.com/document/d/17MGIkufLUZW5XnFoU9eRcMZ6qQ5-DJnSs-wwjfApig8/edit) was just shared with me. Why do you claim as part of the proposal “We haven’t discussed this project and the proposal in any channel before. “ and the proposal itself contains funding for the “Game Center Module”? This is the exact same module you want to get funded via grants. 
Also I still believe it doesn’t make so much sense to put the “Substrate Module: Distribution HPT” on-chain. This can probably be archived much simpler with an off-chain module, since the module itself would only be “live” once you parachain actually is up and running. https://www.kilt.io/polimec/ tries to solve this in a more general way.

**Comment by HexPlayLab** [comment]
> Thanks for the update. Your [treasury proposal](https://docs.google.com/document/d/17MGIkufLUZW5XnFoU9eRcMZ6qQ5-DJnSs-wwjfApig8/edit) was just shared with me. Why do you claim as part of the proposal “We haven’t discussed this project and the proposal in any channel before. “ and the proposal itself contains funding for the “Game Center Module”? This is the exact same module you want to get funded via grants.
> Also I still believe it doesn’t make so much sense to put the “Substrate Module: Distribution HPT” on-chain. This can probably be archived much simpler off-chain, since the module itself would only be “live” once you parachain actually is up and running. https://www.kilt.io/polimec/ tries to solve this in a more general way.

In that propose, I may have a wrong interpretation of the word "channel". I thought it is a room in element, since the [Treasury Proposal spending](https://docs.google.com/document/d/1IZykdp2cyQavcRyZd_dgNj5DcgxgZR6kAqGdcNARu1w/edit) uses the sentence "To access the channels please create a Riot/Element account if you do not have one yet.", I thought channel referred to a room/channel in element. Because I never discussed that project in element, so I said no discussion in any channel. 

The project I submitted for polkadot treasury and web3 foundation is actually the same one. I saw the polkadot treasury option in the "alternative funding" section in the web3 grant GitHub webpage, and didn't see any restriction that the same project can't apply for polkadot treasury. Therefore, I submitted the polkadot treasury spending proposal.

If there is any misunderstanding, I am happy to explain.

For the part of distribution HPT module, our team may need some time to do some research before giving a response.

**Comment by HexPlayLab** [comment]
> Thanks for the update. Your [treasury proposal](https://docs.google.com/document/d/17MGIkufLUZW5XnFoU9eRcMZ6qQ5-DJnSs-wwjfApig8/edit) was just shared with me. Why do you claim as part of the proposal “We haven’t discussed this project and the proposal in any channel before. “ and the proposal itself contains funding for the “Game Center Module”? This is the exact same module you want to get funded via grants.
> Also I still believe it doesn’t make so much sense to put the “Substrate Module: Distribution HPT” on-chain. This can probably be archived much simpler off-chain, since the module itself would only be “live” once you parachain actually is up and running. https://www.kilt.io/polimec/ tries to solve this in a more general way.

Regarding the functionality of distributing native token, https://www.kilt.io/polimec/ is really what we need. What we actually need to implement is to design a scheme to distribute native token and feed it into the off-chain application provided by KILT.
Therefore, it's not actually a module, and we remove the "distribution module" from the milestone.

File: applications/HexPlay.md [hunk 1]
```diff
@@ -157,7 +157,7 @@ The primary back-end development will be implemented in rust programming languag
 | 6.     | Testing                | The code will have proper unit-test coverage to ensure functionality and robustness. |
 | 7.     | Docker                 | We will provide a docker file to demonstrate the functionality of our chain with the corresponding module. |
 
-### Mildestone 2: Staking-Mining module, Distribution Module, and their Front-End
+### Mildestone 2: Staking-Mining module, and their Front-End
 * **Estimated Duration:** 1 months
 * **FTE:**  4
 * **Costs:** 0.3 BTC
```
Comment (mmagician) [inline applications/HexPlay.md:163]: That's quite expensive for one module that now forms this milestone. Can you at least halve it please?

Comment (HexPlayLab) [reply applications/HexPlay.md:163]: We are thinking about merger M1 and M2, and require 2 month of time, and a total of 0.5 BTC. will this look good? Do you have any idea on how to group the work together? Thanks a lot.

Comment (mmagician) [reply applications/HexPlay.md:163]: Merging sounds reasonable to me.

> Do you have any idea on how to group the work together? 

What do you mean by this? You can edit the application file and re-arrange the deliverables from M2 to M1, thus effectively deleting M2.

**Comment by mmagician** [comment]
Could you also perhaps comment on the similarities to [this application](https://github.com/w3f/Open-Grants-Program/pull/134) please?

Finally, I would also like to know how "a game" is defined:
- what information is submitted by the users upon creation via the substrate module?
- how does this information translate to a playable game?
- where is the game execution logic defined (valid moves etc.)
- where is it executed (on chain? your platform?)

**Comment by HexPlayLab** [comment]
> Could you also perhaps comment on the similarities to [this application](https://github.com/w3f/Open-Grants-Program/pull/134) please?
> 
> Finally, I would also like to know how "a game" is defined:
> 
> * what information is submitted by the users upon creation via the substrate module?
> * how does this information translate to a playable game?
> * where is the game execution logic defined (valid moves etc.)
> * where is it executed (on chain? your platform?)

1. Regarding the part of the game, PR#134 require using the templates to create the game, but in our case, it really depends on what the community wants. In a sense, we have less restriction on the creation of the game. In terms of the economic model, we didn't see a concrete economic model provided in PR#134. HexPlay uses the way of locking HPT/LP Token to mine the profit along with the role of the treasury to help the platform work in a self-sustainable way.
2. First of all, HexPlay will have some general rules for the game to be added on the platform, such as a game must be profitable enough to support itself (the online transaction fee), how much on-chain resources could be used, etc. Those rules are to be designed in the future. The exact implementation of the game, which part to put on-chain, or which part to put off-chain; those things depends on the develop team and the exact game instance. Finally, the community will have a vote to decide if they are satisfied with the design and implementation of the game.
3. M1 and M2 are merged.

**Comment by mmagician** [comment]
>  PR#134 require using the templates to create the game, but in our case, it really depends on what the community wants.

So you don't yet know how game creation will work? In your application, you say: `The entire substrate game module will be completed, including more operations like update a game, remove a game and add a new game.`
Building a community takes time. If you want to first develop an understanding of what your users want, I suggest this should be done prior to applying for a grant. In this grant, we expect you to present concrete solutions. In this case, it seems that you don't have the process of creating new games fleshed out.

Furthermore:
> such as a game must be profitable enough to support itself (the online transaction fee)

Who is going to check this, and how?

Perhaps to make everything clearer, you could describe how your CryptoZoo game will fit into this picture (e.g. how is the game created using your `Game Center` module, what is stored on/off chain, how do users interact etc. ).

Please update the application with that information.

**Comment by mmagician** [comment]
Closing due to inactivity. Feel free to re-open the application with a more complete proposal.

**Pull request closed without merging**